syntax = "proto3";

package message_types;

message Body {
	string type = 1;
	bytes data = 2;
}

message Handshake {
	int64 msg_seq = 1;
	string sender_id = 2;
	string sender_name = 3;
	int64 timestamp = 4;
}

message Heartbeat {
	string sender_id = 1;
	int64 timestamp = 2;
}

message Proposal {
	string msg_id = 1;
	string sender_id = 2;
	repeated string prev_sender_ids = 3;
	string topic_id = 4;
	int64 topic = 5;
	int64 timestamp = 6;
}

message Vote {
	string msg_id = 1;
	string sender_id = 2;
	repeated string next_receiver_ids = 3;
	string topic_id = 4;
	bool value = 5;
	int64 timestamp = 6;
}

message Result {
	string msg_id = 1;
	int64 msg_seq = 2;
	string sender_id = 3;
	repeated Body bodies = 6;
	int64 timestamp = 7;
}

message Message {
	oneof msg_type {
		Heartbeat heartbeat = 2;
		Proposal proposal = 3;
		Vote vote = 4;
		Result result = 5;
	}
}
